Document for using Azure external dns with OpenShift. (#71)
openshift/external-dns-operator: docs/azure-openshift.md

$ RESOURCE_GROUP=$(oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_resourcegroup}} | base64 -d)
$ SUBSCRIPTION_ID=$(oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_subscription_id}} | base64 -d)
$ TENANT_ID=$(oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_tenant_id}} | base64 -d)
```

4. Login to azure with base64 decoded values you get from above

```bash
$ az login --service-principal -u "${CLIENT_ID}" -p "${CLIENT_SECRET}" --tenant "${TENANT_ID}"
```

5. Get the routes to check the domain
```bash
oc get routes --all-namespaces | grep console
openshift-console          console             console-openshift-console.apps.test-azure.qe.azure.devcluster.openshift.com                       console             https   reencrypt/Redirect     None
openshift-console          downloads           downloads-openshift-console.apps.test-azure.qe.azure.devcluster.openshift.com                     downloads           http    edge/Redirect          None
```

6. Get the dns zone list w.r.t your resource group.
```bash

> $ $ az network dns zone list --resource-group "${RESOURCE_GROUP}"

Reviewer: ```suggestion
$ az network dns zone list --resource-group "${RESOURCE_GROUP}"
```

[]
```
Initially you will see nothing. So you will have to create a zone for your resource group.

7. Create a dns zone for you resourcegroup.
```bash
$ az network dns zone create -g "${RESOURCE_GROUP}" -n <pick up the zone name from the route  eg- the part after apps. eg - test-azure.qe.azure.devcluster.openshift.com>
```

8. Then create CR as follows -  
```yaml
$ cat <<EOF | oc create -f -
apiVersion: externaldns.olm.openshift.io/v1alpha1
kind: ExternalDNS
metadata:
  name: sample-azure
spec:
  domains:
  - filterType: Include
    matchType: Exact
    name: test-azure1.qe.azure.devcluster.openshift.com
  provider:
    type: Azure
  source:

>     # Incase you have multiple ingress controllers you must specify ingress controller name in the routerName
>     # so that the external dns will use the router canonical name correrponding to it to create a dns record.
>     openshiftRouteOptions:
>       routerName: default
>     #Source Type is route resource of OpenShift

Reviewer: ```suggestion
    openshiftRouteOptions:
      routerName: default
```
I think that comments can be removed to save some space. BTW you can link the examples from `config/samples/azure/*` to show different options.

    fqdnTemplate:
    - "{{.Name}}.apps.test-azure1.qe.azure.devcluster.openshift.com"
    type: OpenShiftRoute
  zones:
  - "/subscriptions/53b4f551-f0fc-4bea-8cba-11111111111/resourceGroups/test-azure1-nxkxm-rg/providers/Microsoft.Network/dnszones/test-azure1.qe.azure.devcluster.openshift.com"
EOF
```

9. Now you shall see records getting created for OCP created routes using the following command -
```bash
$ az network dns record-set list -g "${RESOURCE_GROUP}"  -z <zone name you created in step 7 eg - test-azure.qe.azure.devcluster.openshift.com>  > record-list-azure.txt
```

10. You can try to create a route with a sample app.
```bash
$ oc new-app --docker-image=openshift/hello-openshift -l app=hello-openshift
$ oc expose service/hello-openshift -l app=hello-openshift
```

11. `hello-openshift` DNS records shall exist in the zone:
```bash
$ az network dns record-set list -g "${RESOURCE_GROUP}" -z <zone name you created in step 7 eg - test-azure.qe.azure.devcluster.openshift.com>  | grep -c "hello-openshift"
```

13. Similarly you can delete the route and app
```bash

> oc delete -f <resources>.yaml

Reviewer: ```suggestion
$ oc delete all -l  app=hello-openshift
```

openshift-console          downloads           downloads-openshift-console.apps.misalunk-azure.qe.azure.devcluster.openshift.com                     downloads           http    edge/Redirect          None
```

6. Get the dns zone list w.r.t your azure_resourcegroup. Get it from the base64 decoded output.
```bash
$ az network dns zone list --resource-group <azure_resourcegroup>
[]
```
Initially you will see nothing. So you will have create a zone for your resourcegroup.

7. Create a dns zone for you resourcegroup.
```bash
 az network dns zone create -g <azure_resourcegroup> -n <pick up the zone name from the route  eg- the part after apps. eg - misalunk-azure.qe.azure.devcluster.openshift.com>
```

8. Then create CR as follows -

```yaml
  apiVersion: externaldns.olm.openshift.io/v1alpha1
  kind: ExternalDNS
  metadata:
    name: sample-azure
  spec:
    domains:
    - filterType: Include
      matchType: Exact
      name: misalunk-azure.qe.azure.devcluster.openshift.com
    provider:

>       type: Azure

Reviewer 2: same question, in the doc https://github.com/openshift/external-dns-operator/blob/main/docs/usage.md#azure the `azure.Configfile.name` and a secret are required, so they are not needed any more?

Author: Cco will fetch the credentials in aws, azure and gcp. The code has to me merged yet.

    source:
      hostnameAnnotation: Allow
      openshiftRouteOptions:
        routerName: default

>       type: OpenShiftRoute

Reviewer 2: before I can see the section below for publishing service to external DNS, is that still supported or removed? 
```
  source:
    type: Service
    service:
      serviceType:
        - LoadBalancer
        - ClusterIP
    annotationFilter:
      external-dns.mydomain.org/publish: "yes"
    fqdnTemplate:
      - "{{.Name}}.test-azure.example.com"
```

Author: These are not needed.

Author: BTW I have sent PRs for sample CR's. @alebedev87 has to approve them yet though.

$ oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_resource_prefix}} | base64 -d
$ oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_resourcegroup}} | base64 -d
$ oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_subscription_id}} | base64 -d
$ oc get secrets azure-credentials  -n kube-system  --template={{.data.azure_tenant_id}} | base64 -d

```

4. Login to azure with base64 decoded values you get from above

```bash
$ az login --service-principal -u <azure_client_id> -p <azure_client_secret> --tenant <azure_tenant_id>
```

5. Get the routes to check the domain
```bash
oc get routes --all-namespaces | grep console
openshift-console          console             console-openshift-console.apps.misalunk-azure.qe.azure.devcluster.openshift.com                       console             https   reencrypt/Redirect     None
openshift-console          downloads           downloads-openshift-console.apps.misalunk-azure.qe.azure.devcluster.openshift.com                     downloads           http    edge/Redirect          None
```

> 6. Get the dns zone list w.r.t your azure_resourcegroup. Get it from the base64 decoded output.
> ```bash
> $ az network dns zone list --resource-group <azure_resourcegroup>
> []
> ```

Reviewer: ```suggestion
6. Get the dns zone list w.r.t your resource group.
```bash
$ az network dns zone list --resource-group "${RESOURCE_GROUP}"
[]
```
```

> Initially you will see nothing. So you will have create a zone for your resourcegroup.

Reviewer: ```suggestion
Initially you will see nothing. So you will have to create a zone for your resource group.
```

7. Create a dns zone for you resourcegroup.
```bash

>  az network dns zone create -g <azure_resourcegroup> -n <pick up the zone name from the route  eg- the part after apps. eg - misalunk-azure.qe.azure.devcluster.openshift.com>

Reviewer: ```suggestion
 $ az network dns zone create -g "${RESOURCE_GROUP}" -n <pick up the zone name from the route  eg- the part after apps. eg - misalunk-azure.qe.azure.devcluster.openshift.com>
```

```

> 8. Then create CR as follows -
> ```yaml
> apiVersion: externaldns.olm.openshift.io/v1alpha1
> kind: ExternalDNS
> metadata:
>   name: sample-azure
> spec:
>   domains:
>   - filterType: Include
>     matchType: Exact
>     name: test-azure1.qe.azure.devcluster.openshift.com
>   provider:
>     type: Azure
>   source:
>     # Incase you have multiple ingress controllers you must specify ingress controller name in the routerName
>     # so that the external dns will use the router canonical name correrponding to it to create a dns record.
>     openshiftRouteOptions:
>       routerName: default
>     #Source Type is route resource of OpenShift
>     fqdnTemplate:
>     - "{{.Name}}.apps.test-azure1.qe.azure.devcluster.openshift.com"
>     type: OpenShiftRoute
>   zones:
>   - "/subscriptions/53b4f551-f0fc-4bea-8cba-11111111111/resourceGroups/test-azure1-nxkxm-rg/providers/Microsoft.Network/dnszones/test-azure1.qe.azure.devcluster.openshift.com"
> ```
> 
> 9. Create this CR using 
> ```bash
> oc create -f <cr-file-name>.yaml
> ```

Reviewer: ```suggestion
8. Then create CR as follows -
```yaml
$ cat <<EOF | oc create -f -
apiVersion: externaldns.olm.openshift.io/v1alpha1
kind: ExternalDNS
metadata:
  name: sample-azure
spec:
  domains:
  - filterType: Include
    matchType: Exact
    name: test-azure1.qe.azure.devcluster.openshift.com
  provider:
    type: Azure
  source:
    # Incase you have multiple ingress controllers you must specify ingress controller name in the routerName
    # so that the external dns will use the router canonical name correrponding to it to create a dns record.
    openshiftRouteOptions:
      routerName: default
    #Source Type is route resource of OpenShift
    fqdnTemplate:
    - "{{.Name}}.apps.test-azure1.qe.azure.devcluster.openshift.com"
    type: OpenShiftRoute
  zones:
  - "/subscriptions/53b4f551-f0fc-4bea-8cba-11111111111/resourceGroups/test-azure1-nxkxm-rg/providers/Microsoft.Network/dnszones/test-azure1.qe.azure.devcluster.openshift.com"
EOF
```
```

> 10. Now you shall see records getting created for OCP created routes using the following command -
> ```bash
> az network dns record-set list -g <resourcegroup>  -z <zone name you created in step 7 eg - misalunk-azure.qe.azure.devcluster.openshift.com>  > record-list-azure.txt
> ```

Reviewer: ```suggestion
10. Now you shall see records getting created for OCP routes using the following command:
```bash
$ az network dns record-set list -g "${RESOURCE_GROUP}"  -z <zone name you created in step 7 eg - misalunk-azure.qe.azure.devcluster.openshift.com>  > record-list-azure.txt
```
```

11. You can try to create a route with a sample app.
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: hello-openshift
  name: hello-openshift
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello-openshift
  template:
    metadata:
      labels:
        app: hello-openshift
    spec:
      containers:
      - image: openshift/hello-openshift
        name: hello-openshift
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: hello-openshift
  name: hello-openshift
spec:
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    app: hello-openshift
  sessionAffinity: None
  type: ClusterIP
---
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  name: hello-openshift
spec:
  host: hello-openshift.apps.<your domain, check your routes for reference>
  to:
    kind: Service
    name: hello-openshift
    weight: 100
  wildcardPolicy: None

```

12. Create this route from the above file
```bash
$ oc create -f <resources>.yaml

> ```

Reviewer: Same as here: https://github.com/openshift/external-dns-operator/pull/72#discussion_r767699193

> 13. Get the records for the dns zone
> ```bash
> az network dns record-set list -g <resourcegroup>  -z <zone name you created in step 7 eg - misalunk-azure.qe.azure.devcluster.openshift.com>  > record-list-azure.txt
> ```
> 
> 14.  Search for "hello" in the record-list-azure.txt and it shall exist.

Reviewer: ```suggestion
13. `hello-openshift` DNS records shall exist in the zone:
```bash
$ az network dns record-set list -g "${RESOURCE_GROUP}" -z <zone name you created in step 7 eg - misalunk-azure.qe.azure.devcluster.openshift.com>  | grep -c "hello-openshift"
```

14.  Search for "hello" in the record-list-azure.txt and it shall exist.
```

15. Similarly you can delete the route and app
```bash
oc delete -f <resources>.yaml
```

> 16. Then again run the command to get the records
> ```bash
> az network dns record-set list -g <resourcegroup>  -z <zone name you created in step 7 eg - misalunk-azure.qe.azure.devcluster.openshift.com>  > record-list-azure.txt
> ```
> 
> 17. Search for "hello" in the record-list-azure.txt and it shall not exist.

Reviewer: ```suggestion
16. Then again run the command to check that `hello-openshift` records are not there anymore:
```bash
$ az network dns record-set list -g "${RESOURCE_GROUP}"  -z <zone name you created in step 7 eg - misalunk-azure.qe.azure.devcluster.openshift.com>  | grep -c "hello-openshift"
```
```
